### DISCUSSION TITLE
Screenshot (Utility) UI & UX Overhaul

### DISCUSSION THREAD

**Discussion by cyrneko** [elementary/triage#585] [Ideas]
elementaryOS already has a fine screenshot tool, even with the unique ability of temporarily censoring (system-font) text, which is pretty nifty!

Though, there are also a couple of UI/UX issues with screenshots as they currently are:
- There is no feedback to where screenshots are saved to when using shortcuts
- All shortcuts dismiss menus (especially frustrating when trying to crop to display a context menu)
- Screenshot is only taken upon *finishing* a selection, not upon *pressing the button*
- Without looking at the shortcuts, users may not know other shortcuts apart from only <kbd>print</kbd>, or even other options than taking one full-screen screenshot

## A revamped screenshot UI for the <kbd>print</kbd> shortcut

This is very much ~~stolen~~ inspired by GNOME, but having the default <kbd>print</kbd> keyboard shortcut display an overlay with screenshot options (e.g fullscreen, cropped, conceal text, etc) and defaulting to showing a template rectangle selection actually seems like a really good idea from the GNOME team to convey:
- Other options than fullscreen exist
- Here is how to use the crop tool
- This is what you'll be screenshotting

This is particularly nice because GNOME, in the same release, also made the screenshot be taken immediately and then displayed full-screen in an overlay, together with (behind the scenes) taking a window-screenshot of all open windows. This effectively means that whatever the user selects will be what was on-screen the moment they pressed <kbd>print</kbd>, and NOT the moment they confirmed taking a screenshot. It'll also be easier to take screenshots of menus and other things that would otherwise be dismissed.

If you've not seen the new(-ish) GNOME screenshot UI, here is a screenshot (ironic, I know):
![image](https://github.com/user-attachments/assets/fa4324d9-ff81-4d77-96a2-2da10dbfae85)
[taken from the GNOME Shell Merge Request](https://gitlab.gnome.org/GNOME/gnome-shell/-/merge_requests/1954)

This of course raises the question of **what to do with the existing screenshot tool?**
I think there is generally two ways around this:
- Do it the GNOME way: Depracate the tool, and don't include it in the main desktop anymore. Add an option in the shell's search to take a screenshot (in the elementaryOS case this would mean simply replacing the screenshot GUI tool with a shortcut of the same name that opens the new UI)
- Have the two co-exist, with the GUI behaving exactly the same as it does now, maybe using a compositor-provided backend like the old GNOME Screenshot tool does.

## Out-of-scope:
- editing tools, those belong in [photos](https://github.com/elementary/photos)
- screen recording

## Issues with this approach:

Notably, one thing that this would break, or at least make harder to implement with UX as the user expects it, is the "conceal text" option. This option entirely relies on the screenshot being taken *after* you hit the button to confirm, and simply wouldn't work if you took a screenshot immediately.

I'm not yet sure how to deal with this myself, either :thinking: 

## Checklist

Just a checklist of things that would be useful to progress this forward:
- [ ] Resolve concerns (see 'issues with this approach')
- [ ] Actual design
- [ ] Implementation
- [ ] Actually test the thing
- [ ] Yay, release :)

Ofc this does not represent expectations of this proposal to ever succeed, but sets a baseline of expectations for rough goals to land this in elementaryOS.

**Comment by darltrash** [comment]
I am still wondering exactly how could this be implemented, usually Pantheon implements all of it's elements as separate processes from the actual compositor, but I am not sure exactly how could this feature be implemented as it's own program.

What if <kbd>print</kbd> simply just launched the screenshots app? 

The screenshots app could take the picture instantly on being opened up, to then show up some UI for cropping the picture. 
The issue that comes with this approach is the fact that it would make implementing the conceal text feature much harder to implement, and require some stuff like some OCR program (which would in fact also have it's own benefits, such as censoring hopefully ALL text on the screen, not just apps using GTK3 and GTK4)

**Reply by darltrash** [reply]
I am prototyping this out in Vala as we speak.

**Reply by cyrneko** [reply]
> but I am not sure exactly how could this feature be implemented as it's own program.

There are programs that do this, personally I've installed Flameshot and bound it to <kbd>print</kbd> as to get a decent chunk of this functionality.

**Reply by darltrash** [reply]
> > but I am not sure exactly how could this feature be implemented as it's own program.
> 
> There are programs that do this, personally I've installed Flameshot and bound it to print as to get a decent chunk of this functionality.

Hm, I still feel like making something specific to elementaryOS, following it's own design principles, would be much better to have

**Reply by cyrneko** [reply]
I would agree, thats why I wrote all this :)
